### PR TITLE
Apigov 18438 - add APIService deletion validation

### PR DESCRIPTION
**Pull request by dfeldick**
Pay special attention to index.md, the SDK documentation. RegisterAPIValidator was not previously documented, so I did that. And then I documented the new RegisterDeleteServiceValidator. Note that I also fixed capitalizations of AMPLIFY, fixed a few grammatical things. Note that the word "setup" is a noun. In most cases here, it is used as a verb so it is correctly "set up".

### CODE REVIEW

File: errors.md [hunk 1]
```diff
@@ -38,6 +38,8 @@
 | 1157 | error getting catalog item API server info properties                                                       | pkg/apic/ErrGetCatalogItemServerInfoProperties      |
 | 1158 | subscription manager is not in a running state                                                              | pkg/apic/ErrSubscriptionManagerDown                 |
 | 1160 | error getting endpoints for the API specification                                                           | pkg/apic/ErrSetSpecEndPoints                        |
+| 1161 | error deleting API Service for catalog item in AMPLIFY Central                                              | pkg/agent/ErrDeletingService                         |
+| 1162 | error deleting catalog item in AMPLIFY Central                                                              | pkg/agent/ErrDeletingCatalogItem                     |
```
Comment (jcollins-axway) [inline errors.md:42]: Since updating everywhere else, Amplify?

File: pkg/agent/discoverycache.go [hunk 2]
```diff
@@ -186,13 +187,37 @@ func validateAPIOnDataplane(consumerInstances []apiV1.ResourceInstance) {
...
+	if agent.deleteServiceValidator == nil {
+		return true
+	}
+	// let the agent decide if service should be deleted
+	return agent.deleteServiceValidator(apiID, stage)
+}
+
+func deleteConsumerInstanceOrService(consumerInstance *v1alpha1.ConsumerInstance, externalAPIID, externalAPIStage string) {
+	if shouldDeleteService(externalAPIID, externalAPIStage) {
+		log.Infof("API no longer exists on the dataplane; deleting the API Service and corresponding catalog item %s from AMPLIFY Central", consumerInstance.Title)
```
Comment (jcollins-axway) [inline pkg/agent/discoverycache.go:206]: Amplify?

File: pkg/apic/apiservice_test.go [hunk 3]
```diff
@@ -285,16 +316,68 @@ func TestUpdateService(t *testing.T) {
...
 	client, httpClient := GetTestServiceClient()
 	httpClient.ResponseCode = http.StatusRequestTimeout
-	err := client.deleteConsumerInstance("12345")
+	err := client.DeleteConsumerInstance("12345")
 	assert.NotNil(t, err)
 	assert.Contains(t, "[Error Code 1120] - error making a request to AMPLIFY: status - 408", err.Error())
 
 	httpClient.ResponseCode = http.StatusNoContent
-	err = client.deleteConsumerInstance("12345")
+	err = client.DeleteConsumerInstance("12345")
```
Comment (jcollins-axway) [inline pkg/apic/apiservice_test.go:324]: Changing this from private to public may cause test failures in agents

File: errors.md (hunk 1 above)
Comment (dfeldick) [reply errors.md:42]: Found 10 others in there, also. Fixed.

File: pkg/agent/discoverycache.go (hunk 2 above)
Comment (dfeldick) [reply pkg/agent/discoverycache.go:206]: done

File: pkg/apic/testdata/consumerinstanceemptylist.json [hunk 4]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (jcollins-axway) [inline pkg/apic/testdata/consumerinstanceemptylist.json:1]: if both files are empty lists, why not just use a single empty list file?

File: pkg/cache/cache.go [hunk 5]
```diff
@@ -140,11 +140,15 @@ func (c *itemCache) set(key string, data interface{}) error {
...
+	secKeys := make(map[string]bool)
+	if _, ok := c.Items[key]; ok {
+		secKeys = c.Items[key].SecondaryKeys
+	}
 	c.Items[key] = &Item{
 		Object:        data,
 		UpdateTime:    time.Now().Unix(),
 		Hash:          hash,
-		SecondaryKeys: make(map[string]bool),
+		SecondaryKeys: secKeys,
```
Comment (jcollins-axway) [inline pkg/cache/cache.go:151]: Good catch

File: pkg/agent/discoverycache.go (hunk 2 above)
Comment (jcollins-axway) [reply pkg/agent/discoverycache.go]: you get the other places in that method as well?

**Pull request merged**